support different scale/zero data types (int8, bf16, fp16) for mixed input mma

taozha2 opened the pull request.

aacostadiaz commented on examples/sycl/09_bmg_gemm_int4/09_bmg_gemm_int4.cpp:
```diff
@@ -0,0 +1,713 @@
+ *
+ **************************************************************************************************/
+/***************************************
+* Mixed Precision PVC Gemm Example
+*
+* This example demonstrates how to dispatch a mixed precision GEMM on PVC, with optional dequantization.
+* The GemmMode enum describes the 3 modes of operation:
```
Replace PVC with BMG

aacostadiaz commented on examples/sycl/09_bmg_gemm_int4/09_bmg_gemm_int4.cpp:
```diff
@@ -0,0 +1,713 @@
+
+#define CACHE_CNT (2)
+
+using MmaType = _Float16;
```
use cutlass::half_t instead

aacostadiaz commented on examples/sycl/09_bmg_gemm_int4/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+set(TEST_BATCHES --l=2)
+
+cutlass_example_add_executable(
+  09_bmg_gemm_int4
+  09_bmg_gemm_int4.cpp
+  TEST_COMMAND_OPTIONS
+  TEST_BATCHES
+)
```
```suggestion
set(TEST_BATCHES --l=2)
set(TEST_MODE_0 --l=2 --mode=0)
set(TEST_MODE_1 --l=2 --mode=1)
set(TEST_CHANNELWISE --l=2 --k= 4096 --g=4096 )

cutlass_example_add_executable(
  09_bmg_gemm_int4
  09_bmg_gemm_int4.cpp
  TEST_COMMAND_OPTIONS
  TEST_BATCHES
)
```

That will run all these configurations on CI

t4c1 commented on examples/sycl/02_bmg_gemm_mixed_dtype/CMakeLists.txt:
```diff
@@ -32,15 +32,28 @@ set(TEST_MODE_0 --l=2 --mode=0)
+  02_bmg_gemm_f16_u4
+  02_bmg_gemm_f16_u4.cpp
+  TEST_COMMAND_OPTIONS
+  # TODO: have bugs for below command, will fix later
+  #TEST_BATCHES
+  #TEST_MODE_1
+  #TEST_MODE_0
+  #TEST_A_NARROW
```
There is a todo left in the code. Either fix it or change it to "TODO(Codeplay): " to be consistent with the others we added if you intend it to be merged.

taozha2 replied:
I think this point doesn't block merging this PR. because of:
1. TEST_BATCHES, TEST_MODE_1 and TEST_MODE_0 are int4 bugs, i have fixed them.
2. TEST_A_NARROW case is a feature for int4 which isn't a bug or regression.
3. The original bf16 x int8_t case can work with its all command case like, i never comments any test command of it.
            cutlass_example_add_executable(
              02_bmg_gemm_bf16_s8
              02_bmg_gemm_bf16_s8.cpp
              TEST_COMMAND_OPTIONS
              TEST_BATCHES
              TEST_MODE_1
              TEST_MODE_0
              TEST_A_NARROW
            )

t4c1 commented on include/cute/arch/copy_xe_U4.hpp:
```diff
@@ -36,6 +36,22 @@
 
 namespace cute
 {
+struct XE_2D_U4x8x16_LD_T {
```
Can you add tests for the new copy atoms in /test/unit/cute/intel_xe/copy_block.cpp?

t4c1 commented on include/cute/atom/copy_traits_xe.hpp:
```diff
@@ -246,7 +246,7 @@ CUTE_HOST_DEVICE auto prefetch_selector(Tensor const& tensor) {
   using PrefetchTraits = Copy_Traits<PrefetchOp, decltype(tensor.stride())>;
   using PrefetchAtom = Copy_Atom<PrefetchTraits, dtype>;
-  using Scalar = Int<cute::max(1, 8 / dtype_size_bits)>;
+  using Scalar = Int<cute::max(1, detail::size_of_inst_bits<typename PrefetchTraits::CopyOp, int8_t> / dtype_size_bits)>;
```
```suggestion
  using Scalar = Int<cute::max(1, detail::size_of_inst_bits<PrefetchOp, int8_t> / dtype_size_bits)>;
```
I think this is equivalent, right?

t4c1 commented on include/cute/atom/copy_traits_xe.hpp:
```diff
@@ -701,6 +719,24 @@ struct Copy_Traits_<XE_2D_U4x16x64_LD_N, args_t...>
+  using SrcLayout = Layout<Shape <_16,Shape <_4, _8>>,
+                           Stride<_0,Stride< _1,_64>>>;
+  // Map from (dst-thr,dst-val) to bit
+  using DstLayout = Layout<Shape <_16,Shape <_4, _8>>,
+                           Stride<_16,Stride< _1,_64>>>;
```
```suggestion
  using DstLayout = Layout<Shape <_16,Shape <_4, _4, _8>>,
                           Stride<_128,Stride< _1,_4,_16>>>;
```

t4c1 commented on include/cute/atom/copy_traits_xe.hpp:
```diff
@@ -736,6 +772,24 @@ struct Copy_Traits_<XE_2D_U4x32x16_LD_T, args_t...>
+                           Stride< _0,_1>>;
+  // Map from (dst-thr,dst-val) to bit
+  using DstLayout = Layout<Shape <_16,_8>,
+                           Stride<_16, _1>>;
```
```suggestion
                           Stride<_8, _1>>;
```

t4c1 commented on include/cute/atom/copy_traits_xe.hpp:
```diff
@@ -557,6 +557,24 @@ struct Copy_Traits_<XE_2D_Packed_U8x1x32_LD_N, args_t...>
+    : XE_2D_LD_Unpack<XE_2D_U8x1x32_LD_N, args_t...> {
+  using ThrID = Layout<_16>;
+  // Map from (src-thr,src-val) to bit
+  using SrcLayout = Layout<Shape <_16,_16>,
+                           Stride< _0,_1>>;
+  // Map from (dst-thr,dst-val) to bit
+  using DstLayout = Layout<Shape <_16,_16>,
+                           Stride<_16, _1>>;
```
You are setting the layouts the same as above one - is this the same operation?

I am guessing you want this here:
```suggestion
  using SrcLayout = Layout<Shape <_16,Shape<_8,_2>>,
                           Stride< _0,Stride<_1,_128>>>;
  // Map from (dst-thr,dst-val) to bit
  using DstLayout = Layout<Shape <_16,Shape<_8,_2>>,
                           Stride<_8, Stride<_1,_128>>>;
```

taozha2 replied:
yes, because this layout are same semantics with above one, so i don't change it. your change is more accurate。

t4c1 commented on include/cute/atom/copy_traits_xe.hpp:
```diff
@@ -2438,6 +2494,7 @@ COPY_TRAIT_LD_DEF(XE_2D_TF32x32x16_LD_N)
 COPY_TRAIT_LD_DEF(XE_2D_U4x32x64_LD_N)
 COPY_TRAIT_LD_DEF(XE_2D_U4x16x64_LD_N)
 COPY_TRAIT_LD_DEF(XE_2D_U4x32x16_LD_T)
+COPY_TRAIT_LD_DEF(XE_2D_U4x8x16_LD_T)
```
You did not add copy traits for the new prefetch atoms.

taozha2 replied:
i think we don't use LD_T for prefetch, or add as needed later.

t4c1 replied:
No, we do not. I commented here just because it is the closest to where the change should be. 

You added a bunch of prefetch atoms, for example `XE_2D_U8x1x16_LD_N::PREFETCH`. They need copy traits.

t4c1 replied:
For example:
```
template <class... args_t>
struct Copy_Traits_<XE_2D_U8x1x16_LD_N::PREFETCH, args_t...>
    : XE_2D_LD_Unpack<XE_2D_U8x1x16_LD_N::PREFETCH, args_t...> {
  // Logical thread id to thread idx
  using ThrID = Layout<_16>;
  // Map from (src-thr,src-val) to bit
  using SrcLayout = Layout<Shape <_16,_8>,
                           Stride< _0,_1>>;
  // Map from (dst-thr,dst-val) to bit
  using DstLayout = Layout<Shape <_16,_8>,
                           Stride<_8, _1>>;
  // Reference map from (thr,val) to bit
  using RefLayout = DstLayout;
  using CopyInternalType = cute::intel::ushort;
  using XE_2D_LD_Unpack<XE_2D_U8x1x16_LD_N::PREFETCH, args_t...>::XE_2D_LD_Unpack;
};
```
and
```
COPY_TRAIT_LD_DEF(XE_2D_U8x1x16_LD_N::PREFETCH)
```

t4c1 replied:
@taozha2 Please do not resolve conversations with unresolved comments!. This has not been addressed. In fact I would prefer if you leave it to me to resolve conversations I started once I think they have been adaquatly addressed.

t4c1 replied:
Hmm, are you saying we do not actually need prefetch traits? I just checked it and you are right. So this is fine.

t4c1 commented on include/cutlass/gemm/collective/xe_mma_mixed_input.hpp:
```diff
@@ -282,6 +313,101 @@ struct CollectiveMma<
+
+    // for tuning performance
+    static constexpr auto vec_size = scalar;
+    static constexpr auto spilits = loop_cnt / vec_size;
```
Is there a typo in this variable name?

t4c1 commented on include/cutlass/gemm/collective/xe_mma_mixed_input.hpp:
```diff
@@ -282,6 +313,101 @@ struct CollectiveMma<
+        CUTLASS_PRAGMA_UNROLL
+        for (int i = 0; i < vec_size; i++) {
+          auto data = [&](){
+            static_assert(!is_signed<SrcType>::value && !IsATransformed);
+              if constexpr (is_signed<SrcType>::value) {
```
Why is there an if on the condition that is checked by assert just above?

taozha2 replied:
it's a mistake, removed.

taozha2 commented on include/cute/arch/copy_xe_U4.hpp:
```diff
@@ -36,6 +36,22 @@
+                                    int height, int pitch, intel::coord_t coord,
+                                    T *dst) {
+#if defined(CUTE_ARCH_COPY_XE_ENABLED)
+    detail::XeSubgroup2DBlockLoadTranspose<4, 1, 16, 1>{}(baseoffset, width, height, pitch, coord, dst);
```
This is transpose, i think it's 4, 1, 16, 1.

taozha2 commented on examples/common/sycl_common.hpp:
```diff
@@ -42,15 +42,12 @@ bool initialize_block(Element* block, std::size_t size, uint64_t seed=2023) {
-    scope_max = Element(2);
-    scope_min = Element(-2);
+  if (cute::is_signed<Element>::value) {
+    scope_max = Element((2 ^ bits_input) / 2  - 1);
```
thank you. yes, your assumption is right, i have updated, please review it.

t4c1 commented on examples/common/sycl_common.hpp:
```diff
@@ -39,19 +39,8 @@
-    scope_max = Element(8);
-    scope_min = Element(-8);
-  }
+  Element scope_max = std::numeric_limits<Element>::max();
+  Element scope_min = std::numeric_limits<Element>::lowest();
```
Something like this will properly limit the range to a fraction of available to avoid overflow:
```suggestion
  Element scope_max = Element(1 << cute::ceil_div(std::numeric_limits<Element>::digits(), 4));
  Element scope_min = cute::is_signed<Element>::value ? -scope_max : 0;
```

taozha2 commented on examples/sycl/02_bmg_gemm_mixed_dtype/02_bmg_gemm_f16_u4.cpp:
```diff
@@ -0,0 +1,599 @@
+};
+
+// Factory structs to factor out boilerplate code
+namespace helpers{
```
I just copy this part from original example, we can refine them together later, current PR mainly focus on int4 feature.

t4c1 replied:
OK, please do refine them after this PR is merged.

taozha2 commented on include/cutlass/gemm/collective/xe_mma_mixed_input.hpp:
```diff
@@ -525,23 +669,34 @@ struct CollectiveMma<
+
+      if(prefetch_k < k_tile_count) {
+        prefetch(tiled_prefetch_a, pAgA(_,_,_,prefetch_k));
+        prefetch(tiled_prefetch_b, pBgB(_,_,_,prefetch_k));
```
this is old code, have removed in latest code

taozha2 commented on examples/sycl/02_bmg_gemm_mixed_dtype/02_bmg_gemm_bf16_s8.cpp:
```diff
@@ -215,7 +215,7 @@ struct ExampleRunner {
   // Scale and Zero share a stride since the layout and shapes must be the same.
   using StrideScale = typename CollectiveMainloop::StrideScale;
-  using StrideZero = StrideScale; 
+  using StrideZero = typename CollectiveMainloop::StrideZero; 
```
this is framework requirement for int4, i will double confirm this via email and add you i  the mail list.

taozha2 replied:
updated

taozha2 commented on tools/util/include/cutlass/util/mixed_dtype_utils.hpp:
```diff
@@ -182,16 +192,22 @@ static void dequantize(DequantizedElement* dq_buffer,
   auto scale_stride_bcast = make_stride(scale_stride0, make_stride(0, scale_stride1), scale_stride2);
   auto scale_layout_bcast = make_layout(scale_shape_bcast, scale_stride_bcast);
 
+  const auto zero_stride0 = get<0>(stride(zero_layout));
+  const auto zero_stride1 = get<1>(stride(zero_layout));
+  const auto zero_stride2 = get<2>(stride(zero_layout));
+  auto zero_shape_bcast = make_shape(num_rows, make_shape(group_size, scale_k), batches);
+  auto zero_stride_bcast = make_stride(zero_stride0, make_stride(0, zero_stride1), zero_stride2);
+  auto zero_layout_bcast = make_layout(zero_shape_bcast, zero_stride_bcast);
```
updated

taozha2 commented:
@aacostadiaz the nightly build failed due to "undefined reference to `__internal_intel_sub_group_2d_block_read_8b_1r16x1c_cache_controls'" which is issue of compatibility of compiler version i think.
the release tests passed, How to deal with this problem？

aacostadiaz commented:
> @aacostadiaz the nightly build failed due to "undefined reference to `__internal_intel_sub_group_2d_block_read_8b_1r16x1c_cache_controls'" which is issue of compatibility of compiler version i think. the release tests passed, How to deal with this problem？

It is because `__builtin_IB_subgroup_block_read_prefetch_u8_m1k16v1` is not available in the spirv copy functions. You need to add it in include/cute/arch/copy_xe_spirv.hpp, similar of what you did for the prefetch builtin.

taozha2 commented:
> > @aacostadiaz the nightly build failed due to "undefined reference to `__internal_intel_sub_group_2d_block_read_8b_1r16x1c_cache_controls'" which is issue of compatibility of compiler version i think. the release tests passed, How to deal with this problem？
> 
> It is because `__builtin_IB_subgroup_block_read_prefetch_u8_m1k16v1` is not available in the spirv copy functions. You need to add it in include/cute/arch/copy_xe_spirv.hpp, similar of what you did for the prefetch builtin.

@aacostadiaz i have add the intrinsic in spirv, but build failed at copy_block.cpp which seems doesn't related to my changes, can you help to check that?

t4c1 commented on include/cute/atom/copy_traits_xe.hpp:
```diff
@@ -364,8 +369,6 @@ struct XE_2D_LD_Unpack {
     static_assert(is_rmem<TD>::value);
     static_assert(size(SLayout{}) * dtype_bits == size<1>(typename Traits_LD_t::SrcLayout{}),
                   "Src tensor size does not match copy atom size.");
-    static_assert(size(DLayout{}) * dtype_bits == size<1>(typename Traits_LD_t::DstLayout{}),
-                  "Dst tensor size does not match copy atom size.");
```
Why is this assert removed? If it is failing there is likely something wrong with your copy operation.

t4c1 commented on include/cute/atom/copy_traits_xe.hpp:
```diff
@@ -94,27 +94,27 @@ struct choose_prefetch_for_type {
 template <>
 struct choose_prefetch_for_type<4, 1> {
-  using Prefetch = XE_2D_Packed_U8x1x32_LD_N;
+  using Prefetch = XE_2D_U16x1x32_LD_N;
```
Multiple subgroups in a workgroup can work together to use the whole cache line, right? So a prefetch atom can be smaller than cache line. Can you show how does this change affect performance?

t4c1 replied:
The performance is virtually the same with either prefetch atoms.

t4c1 commented on include/cute/atom/copy_traits_xe.hpp:
```diff
@@ -667,6 +670,24 @@ struct Copy_Traits_<XE_2D_U8x32x32_LD_N, args_t...>
+  using SrcLayout = Layout<Shape <_16,Shape <_4,  _8>>,
+                           Stride< _0,Stride< _1, _4>>>;
+  // Map from (dst-thr,dst-val) to bit
+  using DstLayout = Layout<Shape <_16,Shape <_4,  _8>>,
+                           Stride<_16,Stride< _1, _4>>>;
```
```suggestion
  using DstLayout = Layout<Shape <_16,Shape <_4,  _8>>,
                           Stride<_4,Stride< _1, _64>>>;
```
This looks wrong. Should it be like this, assuming this is not a packed operation?

t4c1 commented on include/cute/atom/copy_traits_xe.hpp:
```diff
@@ -703,6 +724,24 @@ struct Copy_Traits_<XE_2D_U4x16x64_LD_N, args_t...>
+  using SrcLayout = Layout<Shape <_16,Shape <_4, _8>>,
+                           Stride<_0,Stride< _1,_64>>>;
+  // Map from (dst-thr,dst-val) to bit
+  using DstLayout = Layout<Shape <_16,Shape <_4, _8>>,
+                           Stride<_16,Stride< _1,_64>>>;
```
```suggestion
  using DstLayout = Layout<Shape <_16,Shape <_4, _8>>,
                           Stride<_32,Stride< _1,_4>>>;
```

taozha2 commented:
> There are still 2 unaddressed comments - one about performance of changed prefetch options and the other about copy traits for the new prefetch atoms.

some comments are outdated, can you check latest code?
